traits/observers/expression.py

```diff
@@ -135,162 +132,94 @@ def _as_graphs(self):
+        return self._expression.create_graphs(branches=[])
 
-    def _new_with_branches(self, nodes):
-        """ Create a new Expression with a new leaf nodes.
 
-        Parameters
-        ----------
-        nodes : list of IObserver
+class _IExpression(abc.ABC):
+    """ Interface to be implemented for objects to be wrapped in
+    Expression. Such objects are responsible for constructing an ObserverGraph
+    with the given information and context.
 
-        Returns
-        -------
-        new_expression : traits.observers.expression.Expression
-        """
-        expression = self._copy()
-        expression._levels.append(nodes)
-        return expression
+    All these objects are used internally.
+    """
 
-    def _copy(self):
-        """ Return a copy of this expression.
+    def create_graphs(self, branches):
+        """ Return a list of ObserverGraph with the given branches.
 
-        Returns
-        -------
-        new_expression : traits.observers.expression.Expression
+        Parameters
+        ----------
+        branches : list of ObserverGraph
+            Graphs to be used as branches.
         """
-        expression = Expression()
-        expression._levels = self._levels.copy()
-        if self._prior_expression is not None:
-            expression._prior_expression = self._prior_expression._copy()
-        return expression
+        raise NotImplementedError("'create_graphs' must be implemented.")
 
 
-def _create_graphs(expression, graphs=None):
-    """ Create ObserverGraphs from a given expression.
+@_IExpression.register
+class _EmptyExpression:
+    """ Empty expression as a placeholder."""
 
-    Parameters
-    ----------
-    expression : traits.observers.expression.Expression
-    graphs : collection of ObserverGraph
-        Leaf graphs to be added.
-        Needed when this function is called recursively.
+    def create_graphs(self, branches):
+        return []
 
-    Returns
-    -------
-    graphs : list of ObserverGraph
-        New graphs
-    """
-    if graphs is None:
-        graphs = []
 
-    for nodes in expression._levels[::-1]:
-        graphs = [
-            _ObserverGraph(node=node, children=graphs) for node in nodes
-        ]
+@_IExpression.register
+class _SingleObserverExpression:
+    """ Container of Expression for wrapping a single observer.
+    Used internally in this module.
+    """
 
-    if expression._prior_expression is not None:
-        graphs = expression._prior_expression._create_graphs(
-            graphs=graphs,
-        )
-    return graphs
+    def __init__(self, observer):
+        self.observer = observer
 
+    def create_graphs(self, branches):
+        return [
+            _ObserverGraph(node=self.observer, children=branches),
+        ]
 
-# _SeriesExpression and _ParallelExpression share an undeclared interface
-# which require the classes to have implemented ``copy`` and ``_create_graphs``
 
+@_IExpression.register
```

Comment:
Again, nothing does `issubclass` or `isinstance` check. The registeration is unused. The registration is only here for code readability and discovery (e.g. it is easier to see what implement the interface).